refactor(rust): manage build via `BuildFactory` for `Bundler`

hyf0 opened the pull request:
Follow-up of https://github.com/rolldown/rolldown/pull/6877.

This PR makes the native `Bundler`​ integrate with the `Build`​ abstraction too.

hyf0 commented:
* **#6887** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6887?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#6888](https://github.com/rolldown/rolldown/pull/6888) <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6888?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#6889](https://github.com/rolldown/rolldown/pull/6889) <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6889?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#6886** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6886?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6884** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6883** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6883?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6883?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6883).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/rolldown/src/bundler/bundler.rs:
```diff
@@ -0,0 +1,127 @@
+    Ok(())
+  }
+
+  // The rollup always crate a new build at watch mode, it cloud be call multiply times.
```
```suggestion
  // The rollup always crate a new build at watch mode, it could be called multiply times.
```

sapphi-red commented on crates/rolldown/src/bundler/impl_bundler_build.rs:
```diff
@@ -0,0 +1,27 @@
+  #[tracing::instrument(level = "debug", skip_all, parent = &self.session.span)]
+  pub async fn generate(&mut self) -> BuildResult<BundleOutput> {
+    self.create_error_if_closed()?;
+
```
```suggestion
```
(just to align with the `write` method above)

sapphi-red commented on crates/rolldown/src/dev/dev_engine.rs:
```diff
@@ -165,7 +165,7 @@ impl DevEngine {
 
     let bundler = self.build_driver.bundler.lock().await;
     // hyf0 TODO: `get_watch_files` is not a proper API to tell which files should be watched.
```
```suggestion
    // hyf0 TODO: `watch_files` is not a proper API to tell which files should be watched.
```

hyf0 committed:
refactor(rust): manage build via `BuildFactory` for `Bundler` (#6883)

Follow-up of https://github.com/rolldown/rolldown/pull/6877.

This PR makes the native `Bundler`​ integrate with the `Build`​ abstraction too.
